src/test/lib/EventTest.cpp

```diff
@@ -557,6 +584,205 @@ void QuicTestValidateConnectionEvents(uint32_t Test)
+    _In_ MsQuicListener& Listener,
+    _In_ QuicAddr& ServerLocalAddr
+    )
+{
+    TestScopeLogger ScopeLogger(__FUNCTION__);
+
+    MsQuicSettings Settings;
+    Settings.SetMinimumMtu(1280).SetMaximumMtu(1280);
+    Settings.SetCongestionControlAlgorithm(QUIC_CONGESTION_CONTROL_ALGORITHM_BBR);
+    Settings.SetNetStatsEventEnabled(true);
+    MsQuicConfiguration ServerConfiguration(Registration, "MsQuicTest", Settings, ServerSelfSignedCredConfig);
+    TEST_TRUE(ServerConfiguration.IsValid());
+
+    MsQuicConfiguration ClientConfiguration(Registration, "MsQuicTest", Settings, MsQuicCredentialConfig());
+    TEST_TRUE(ClientConfiguration.IsValid());
+
+    ConnValidator Client(
+        new(std::nothrow) ConnEventValidator* [4] {
+            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_DATAGRAM_STATE_CHANGED),
+            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_CONNECTED, QUIC_EVENT_ACTION_SHUTDOWN_CONNECTION),
+            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_SHUTDOWN_COMPLETE),
+            nullptr
+        }
+    );
+    ConnValidator Server(
+        new(std::nothrow) ConnEventValidator* [4] {
+            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_CONNECTED, 0, true),
+            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_SHUTDOWN_INITIATED_BY_PEER),
+            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_SHUTDOWN_COMPLETE),
+            nullptr
+        },
+        ServerConfiguration
+    );
+
+    Client.SetEventsExpectedAtleastOnce(
+        new(std::nothrow) ConnEventValidator* [2] {
+            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_NETWORK_STATISTICS),
+            nullptr
+        }
+    );
+
+    Server.SetEventsExpectedAtleastOnce(
+        new(std::nothrow) ConnEventValidator* [2] {
+            new(std::nothrow) ConnEventValidator(QUIC_CONNECTION_EVENT_NETWORK_STATISTICS),
+            nullptr
+        }
+    );
+
+    Listener.Context = &Server;
+
+    TEST_QUIC_SUCCEEDED(
+        MsQuic->ConnectionOpen(
+            Registration,
+            ConnValidatorCallback,
+            &Client,
+            &Client.Handle));
+    TEST_QUIC_SUCCEEDED(
+        MsQuic->ConnectionStart(
+            Client.Handle,
+            ClientConfiguration,
+            QuicAddrGetFamily(&ServerLocalAddr.SockAddr),
+            QUIC_TEST_LOOPBACK_FOR_AF(
+                QuicAddrGetFamily(&ServerLocalAddr.SockAddr)),
+            ServerLocalAddr.GetPort()));
+
+    TEST_TRUE(Client.Complete.WaitTimeout(2000));
+    TEST_TRUE(Server.Complete.WaitTimeout(1000));
+
+    uint8_t EventIndex = UINT8_MAX;
+    if (!Server.SuccessOfEventsExpectedAtleastOnce(&EventIndex) && EventIndex != UINT8_MAX) {
+        TEST_FAILURE("QuicTestValidateNetStatsEvent1: Expected event did not occur. Event type %u  was expected to happen on server at least once", Server.EventsExpectedAtleastOnce[EventIndex]);
+    }
```

Comment:
I feel like a slightly better design is to push this logic all into `SuccessOfEventsExpectedAtleastOnce`. It loops through all events and if any were missed, you call `TEST_FAILURE`. Then here you only need:
```c
Server.SuccessOfEventsExpectedAtleastOnce();
```

Reply:
Def better. Done